File: stdlib/lua/hull/web/auth-health.lua
--- hull.web.auth-health — runtime health probes for the auth stack.
--
-- @module hull.web.auth-health
-- @license AGPL-3.0-or-later
--
-- A single read-only function `auth_health.check()` that probes the
-- shape of the auth-stack tables and module state. Apps wire it into
-- their existing /ready handler (via hull/web/middleware/health) or
-- a dedicated /admin/auth-status endpoint. The probe is also the
-- backing store for `hull agent auth-status` (the CLI just curls
-- the configured endpoint and pretty-prints the JSON).
--
-- The check does NOT exercise external dependencies (HIBP, IdP
-- JWKS, SMTP). Those would turn a health probe into an availability
-- dependency. Instead we report cached state:
--   * pwned.health() (last attempt status, populated by previous
--     /auth/register attempts)
--   * oauth providers configured (count, no reachability test)
--   * audit_log cleanup scheduled (boolean, from init state)
--   * session table exists + has rows (count)
--   * totp table exists + count of enrolled users
--   * rbac roles + permissions tables exist
--
-- A probe is reported as { ok = bool, ... } where additional fields
-- give specifics. `all_ok` is the AND across every probe.
--
-- ## Usage
--
--     local auth_health = require("hull.web.auth-health")
--     local h = auth_health.check()
--     if not h.all_ok then ... end
--
-- Or wire into hull/web/middleware/health:
--
--     local health      = require("hull.web.middleware.health")
--     local auth_health = require("hull.web.auth-health")
--     health.register("auth", function() return auth_health.check().all_ok end)
--
-- Or expose a JSON endpoint for the agent CLI:
--
--     local auth_health = require("hull.web.auth-health")
--     app.get("/admin/auth-status", function(_req, res)
--         res:json(auth_health.check())
--     end)

local db = require("hull.db")

local auth_health = {}

-- Helper: does a table exist?
local function table_exists(name)
    local rows = db.query(
        "SELECT 1 FROM sqlite_master WHERE type = 'table' AND name = ?",
        { name })
    return rows and #rows > 0 or false
end

-- Helper: how many rows in a table (0 if table missing).
local function row_count(name)
    if not table_exists(name) then return 0 end
    local rows = db.query("SELECT count(*) AS n FROM " .. name)
    return (rows and rows[1] and rows[1].n) or 0
end

-- Probe individual subsystems. Each returns { ok, ...details }.
-- Probes use pcall so a single failure doesn't take down the rest.

local function probe_sessions()
    if not table_exists("_hull_sessions") then
        return { ok = false, reason = "table _hull_sessions not created "
                                   .. "(call session.init())" }
    end
    return { ok = true, sessions = row_count("_hull_sessions") }
end

local function probe_audit_log()
    if not table_exists("_hull_audit_log") then
        return { ok = false, reason = "table _hull_audit_log not created "
                                   .. "(call audit_log.init())" }
    end
    local ok, mod = pcall(require, "hull.web.middleware.audit-log")
    local scheduled = nil
    if ok and mod and type(mod.is_cleanup_scheduled) == "function" then
        scheduled = mod.is_cleanup_scheduled()
    end
    return {
        ok = true,
        events = row_count("_hull_audit_log"),
        cleanup_scheduled = scheduled,
    }
end

local function probe_pwned()
    local ok, mod = pcall(require, "hull.web.pwned")
    if not ok or not mod.health then
        return { ok = false, reason = "hull/web/pwned not available" }
    end
    local h = mod.health()
    return {
        -- ok=true when either the last attempt succeeded OR no attempt
        -- has happened yet (process just started). The fail-open
        -- condition (ok=false) is what operators care about.
        ok            = h.ok ~= false,
        last_check_at = h.last_check_at,
        last_error    = h.last_error,
    }
end

local function probe_totp()
    if not table_exists("_hull_totp") then
        return { ok = false, reason = "table _hull_totp not created "
                                   .. "(call totp.init())" }
    end
    local enrolled = db.query(
        "SELECT count(*) AS n FROM _hull_totp WHERE confirmed = 1")
    return {
        ok = true,
        enrolled_users = (enrolled and enrolled[1] and enrolled[1].n) or 0,
    }
end

local function probe_rbac()
    local roles_ok = table_exists("_hull_roles")
    local perms_ok = table_exists("_hull_permissions")
    if not roles_ok and not perms_ok then
        return { ok = false, reason = "rbac tables not created "
                                   .. "(call rbac.init())" }
    end
    return {
        ok          = roles_ok and perms_ok,
        roles       = row_count("_hull_roles"),
        permissions = row_count("_hull_permissions"),
    }
end

--- Run all probes; return a JSON-ready table.
function auth_health.check()
    local out = {
        sessions  = probe_sessions(),
        audit_log = probe_audit_log(),
        pwned     = probe_pwned(),
        totp      = probe_totp(),
        rbac      = probe_rbac(),
    }
    local all_ok = true
    for _, p in pairs(out) do
        if not p.ok then all_ok = false; break end
    end
    out.all_ok = all_ok
    return out
end

--- Convenience: mount a JSON status endpoint at `/admin/auth-status`
-- (overridable). Backs the `hull agent auth-status` CLI.
-- @tparam table app
-- @tparam[opt] table opts  { path = "/admin/auth-status" }
function auth_health.routes(app, opts)
    opts = opts or {}
    local path = opts.path or "/admin/auth-status"
    app.get(path, function(_req, res) res:json(auth_health.check()) end)
end

return auth_health

File: stdlib/js/hull/web/auth-health.js
/**
 * @file hull:web:auth-health
 * @module hull:web:auth-health
 * @description Runtime health probes for the auth stack.
 *
 * Mirror of stdlib/lua/hull/web/auth-health.lua. See the Lua module
 * header for the design — what's probed, what's deliberately NOT
 * probed (external reachability), and the wiring patterns.
 *
 * @license AGPL-3.0-or-later
 */

import { db as dbModule } from "hull:db";
const db = dbModule.default();
import { auditLog }  from "hull:web:middleware:audit-log";
import { pwned }     from "hull:web:pwned";

// Does a table exist? Uses the dialect-aware column probe (SQLite
// PRAGMA table_info / Postgres information_schema) so the health check
// works on either backend. An empty column set means the table is absent.
function tableExists(name) {
    const cols = db.tableColumns(name);
    return cols && cols.length > 0;
}

function rowCount(name) {
    // All callers pass hardcoded `_hull_*` literals, but this interpolates the
    // name into SQL (SQLite can't parameterize identifiers), so fail closed on
    // anything that isn't a plain identifier -- a future variable table name can
    // never become an injection vector (mirrors auth-health.lua).
    if (typeof name !== "string" || !/^[A-Za-z0-9_]+$/.test(name))
        throw new Error("auth-health.rowCount: table name must be a plain identifier");
    if (!tableExists(name)) return 0;
    const rows = db.query("SELECT count(*) AS n FROM " + name);
    return (rows && rows[0] && rows[0].n) || 0;
}

function probeSessions(includeCounts) {
    if (!tableExists("_hull_sessions")) {
        return { ok: false, reason: "table _hull_sessions not created "
                                 + "(call session.init())" };
    }
    const out = { ok: true };
    if (includeCounts) out.sessions = rowCount("_hull_sessions");
    return out;
}

function probeAuditLog(includeCounts) {
    if (!tableExists("_hull_audit_log")) {
        return { ok: false, reason: "table _hull_audit_log not created "
                                 + "(call auditLog.init())" };
    }
    // Round-11 MEDIUM-7: tri-state cleanup status. See Lua sibling.
    const status = typeof auditLog.cleanupStatus === "function"
        ? auditLog.cleanupStatus() : null;
    const scheduled = typeof auditLog.isCleanupScheduled === "function"
        ? auditLog.isCleanupScheduled() : null;
    const out = { ok: true,
                  cleanup: status,
                  cleanup_scheduled: scheduled };
    if (includeCounts) out.events = rowCount("_hull_audit_log");
    return out;
}

function probePwned() {
    if (!pwned || typeof pwned.health !== "function") {
        return { ok: false, reason: "hull/web/pwned not available" };
    }
    const h = pwned.health();
    // Distinguish "no probe attempted yet" (h.ok == null) from
    // "last probe succeeded" (true) and "last probe failed" (false).
    let status;
    if (h.ok == null)   status = "not_yet_checked";
    else if (h.ok)      status = "reachable";
    else                status = "fail_open";
    return {
        ok:            h.ok !== false,
        status:        status,
        last_check_at: h.last_check_at,
        last_error:    h.last_error,
    };
}

function probeTotp(includeCounts) {
    if (!tableExists("_hull_totp")) {
        return { ok: false, reason: "table _hull_totp not created "
                                 + "(call totp.init())" };
    }
    const out = { ok: true };
    if (includeCounts) {
        const r = db.query(
            "SELECT count(*) AS n FROM _hull_totp WHERE confirmed = 1");
        out.enrolled_users = (r && r[0] && r[0].n) || 0;
    }
    return out;
}

function probeRbac(includeCounts) {
    const rolesOk = tableExists("_hull_roles");
    const permsOk = tableExists("_hull_permissions");
    if (!rolesOk && !permsOk) {
        return { ok: false, reason: "rbac tables not created "
                                 + "(call rbac.init())" };
    }
    const out = { ok: rolesOk && permsOk };
    if (includeCounts) {
        out.roles       = rowCount("_hull_roles");
        out.permissions = rowCount("_hull_permissions");
    }
    return out;
}

/**
 * Run all probes; return a JSON-ready object.
 * opts.includeCounts (default false) gates enumeration counts
 * (events, sessions, enrolled_users, roles, permissions). Default
 * is FALSE — the endpoint surface is often reachable by anyone with
 * admin access and the counts are recon material.
 */
function check(opts) {
    const o = opts || {};
    const include = o.includeCounts === true;
    const out = {
        sessions:  probeSessions(include),
        audit_log: probeAuditLog(include),
        pwned:     probePwned(),
        totp:      probeTotp(include),
        rbac:      probeRbac(include),
    };
    let allOk = true;
    for (const k in out) {
        if (!out[k].ok) { allOk = false; break; }
    }
    out.all_ok = allOk;
    return out;
}

/**
 * Mount /admin/auth-status (overridable). opts.authCheck(req) is
 * REQUIRED — the endpoint exposes session/enrollment counts and
 * operational state, recon material if left open. Must return the
 * boolean `true` to admit; any other value (including truthy ones
 * like `1` or `"yes"`) is rejected with 403; throwing returns 401;
 * returning a Promise (async authCheck) returns 500.
 * opts.includeCounts (default false) gates enumeration counts.
 */
function routes(app, opts) {
    const o = opts || {};
    if (typeof o.authCheck !== "function") {
        throw new Error("authHealth.routes: opts.authCheck function is "
            + "required. The endpoint exposes session/enrollment counts "
            + "and operational state — gate it behind your RBAC "
            + "predicate or a token check.");
    }
    const path = o.path || "/admin/auth-status";
    const includeCounts = o.includeCounts === true;
    const authCheck = o.authCheck;
    app.get(path, (req, res) => {
        let allowed;
        try { allowed = authCheck(req); }
        catch (_e) { return res.status(401).json({ error: "forbidden" }); }
        // Round-8 MEDIUM-5: strict-equality the boolean true. An
        // `async authCheck` returns a Promise, which is truthy under
        // `!allowed` — the gate silently fell open and the handler
        // shipped operational state to anonymous callers. Reject any
        // thenable explicitly so the gate-by-async-mistake fails
        // closed AT the gate, not at the wire.
        if (allowed && typeof allowed.then === "function") {
            return res.status(500).json({
                error: "authCheck must be synchronous; "
                     + "received a thenable (Promise)",
            });
        }
        if (allowed !== true) {
            return res.status(403).json({ error: "forbidden" });
        }
        res.json(check({ includeCounts: includeCounts }));
    });
}

const authHealth = { check, routes };
export { authHealth };
